### ISSUE TITLE
Non-shaped type with >0 results in attn operand

### ISSUE THREAD

**Issue by dan-garvey** [iree-org/iree#21439]
### What happened?

I'm hitting this [assert](https://github.com/iree-org/iree/blob/f87452f3fc7481d820b5c38f6cadbbb6c7f17beb/compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/ReshapeFusion.cpp#L291)

Adding some debug prints left me even more confused:
```
operandType: tensor<1x32x1x256xf16>
num_results: 4
! operandType: 0

operandType: tensor<?x256xf16>
num_results: 2
! operandType: 0

operandType: tensor<?x256xf16>
num_results: 2
! operandType: 0

operandType: f32
num_results: 0
! operandType: 1

operandType: tensor<?x?xf16>
num_results: 2
! operandType: 0

operandType: tensor<1x32x1x256xf32>
num_results: 4
! operandType: 0
```
It seems like I shouldn't be hitting the assert as the only `!operandType` from the prints does have 0 results. What am I missing?

[input ir](https://gist.github.com/dan-garvey/3271194824b4a9b84fb8f4c6b66e0849)

compile command
`iree-compile input.mlir  --iree-hal-target-device=hip  --iree-hip-target=gfx942  -o output.vmfb   --iree-opt-level=O3   --iree-hal-indirect-command-buffers=true   --iree-stream-resource-memory-model=discrete  --iree-hal-memoization=true`

**Comment by IanWood1** [comment]
Something weird is going on, when I ran it with a debug build i'm getting a bad optional access on https://github.com/iree-org/iree/blob/f87452f3fc7481d820b5c38f6cadbbb6c7f17beb/compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/ReshapeFusion.cpp#L417

**Comment by dan-garvey** [comment]
I did some funky stuff with the attention mask that causes it to fail only on decode, when I remove those changes this doesn't occur. Let me double check my math before you spend any more time on this, I'm probably indexing incorrectly

**Comment by dan-garvey** [comment]
updated ir: https://gist.github.com/dan-garvey/3271194824b4a9b84fb8f4c6b66e0849

still hits the same issue.

For some context, the part that is different is the attention mask. I implemented a sliding window via the following python snippet:
```
def sliding_window_only_mask(seq_len, window_left, window_right):
    i = torch.arange(seq_len).unsqueeze(1)  # [T, 1]
    j = torch.arange(seq_len).unsqueeze(0)  # [1, T]
    # Create mask where j is outside [i - left, i + right]
    valid = (j >= (i - window_left)) & (j <= (i + window_right))
    mask = torch.where(valid, torch.tensor(0.0, dtype=torch.float16), torch.tensor(float('-inf'), dtype=torch.float16))
    return mask

if self.attention_window[0] != -1:
    seq_len = xk.size(-2) 
    window_mask = sliding_window_only_mask(seq_len, self.attention_window[0], self.attention_window[1])
    attention_mask = torch.minimum(attention_mask, window_mask)
```

The idea here is that if the window is 1024, and we have a total sequence length of 1034, the mask will have -inf for the first 10 tokens.

**Comment by dan-garvey** [comment]
@Groverkss any ideas? The only thing I can think of is that somewhere we are expecting the mask to be exactly causal and not something else

**Comment by Groverkss** [comment]
> [@Groverkss](https://github.com/Groverkss) any ideas? The only thing I can think of is that somewhere we are expecting the mask to be exactly causal and not something else

No we don't expect that at all

**Comment by IanWood1** [comment]
https://github.com/iree-org/iree/pull/21472 should fix the problem
